core/src/main/java/brooklyn/catalog/internal/CatalogInitialization.java

```diff
@@ -0,0 +1,377 @@
+                if (!catalog.getCatalog().isLoaded()) {
+                    catalog.load();
+                } else {
+                    if (needsInitial && (hasRunOfficial || hasRunBestEffort)) {
+                        // an indication that something caused it to load early; not severe, but unusual
+                        log.warn("Catalog initialization has not properly run but management context has a catalog; re-populating, possibly overwriting items installed during earlier access (it may have been an early web request)");
+                        catalog.reset(ImmutableList.<CatalogItem<?,?>>of());
+                    }
+                }
+                hasRunOfficial = true;
+
+                populateCatalog(catalog, needsInitial, true, optionalItemsForResettingCatalog);
+            }
+        } finally {
+            hasRunOfficial = true;
+            isPopulating = false;
+        }
+    }
+
+    private void populateCatalog(BasicBrooklynCatalog catalog, boolean needsInitial, boolean runCallbacks, Collection<CatalogItem<?, ?>> optionalItemsForResettingCatalog) {
+        applyCatalogLoadMode();
+        
+        if (optionalItemsForResettingCatalog!=null) {
+            catalog.reset(optionalItemsForResettingCatalog);
+        }
+        
+        if (needsInitial) {
+            populateInitial(catalog);
+        }
+        
+        populateAdditions(catalog);
+
+        if (runCallbacks) {
+            populateViaCallbacks(catalog);
+        }
+    }
+
+    private enum PopulateMode { YAML, XML, AUTODETECT }
+    
+    protected void populateInitial(BasicBrooklynCatalog catalog) {
+        if (disallowLocal) {
+            if (!hasRunOfficial()) {
+                log.debug("CLI initial catalog not being read with disallow-local mode set.");
+            }
+            return;
+        }
+
+//        B1) look for --catalog-initial, if so read it, then go to C1
+//        B2) look for BrooklynServerConfig.BROOKLYN_CATALOG_URL, if so, read it, supporting YAML or XML (warning if XML), then go to C1
+//        B3) look for ~/.brooklyn/catalog.bom, if exists, read it then go to C1
+//        B4) look for ~/.brooklyn/brooklyn.xml, if exists, warn, read it then go to C1
+//        B5) read all classpath://brooklyn/default.catalog.bom items, if they exist (and for now they will)
+//        B6) go to C1
+
+        if (initialUri!=null) {
+            populateInitialFromUri(catalog, initialUri, PopulateMode.AUTODETECT);
+            return;
+        }
+        
+        String catalogUrl = managementContext.getConfig().getConfig(BrooklynServerConfig.BROOKLYN_CATALOG_URL);
+        if (Strings.isNonBlank(catalogUrl)) {
+            populateInitialFromUri(catalog, catalogUrl, PopulateMode.AUTODETECT);
+            return;
+        }
+        
+        catalogUrl = Urls.mergePaths(BrooklynServerConfig.getMgmtBaseDir( managementContext.getConfig() ), "catalog.bom");
```

Comment:
If getMgmtBaseDir returns a dir, then clearer to use `Os.mergePaths()`. If it returns a URL, then can't do `new File(catalogUrl).exists()` below, I presume.